### DISCUSSION TITLE
Mock atom inside store for testing

### DISCUSSION THREAD

**Discussion by dephiros** [pmndrs/jotai#1892] [Ideas]
I really love the atom API and with v2 `store`, there are potential for using `jotai` outside of `React`

Is there a way that we can use the store to set a fake atom inside the store so we test without using hook?
```

const atomWithDifficultDep = atom<number>(() => someOtherDep());
const countPlusAtom = atom((get) => {
  return get(atomWithDifficultDep) + 1;
});

// In test
const store = createStore();
const mockedAtom = atom(0);
// some mock API
store.mockAtom(atomWithDifficultDep, mockedAtom);

const values = [];
store.sub((countPlusAtom) => {
  values.push(store.get(countPlusAtom));
});
store.set(mockedAtom, 5);
expect(values.pop()).toEqual(6);
```

**Comment by dai-shi** [comment]
Unfortunately, we don't have a way to override read-only atom values.

(In v1, we had a way. But as it technically breaks data graph in the store, it's prohibited in v2 store.)

You might be able to wrap your store:

```js
const createMockableStore = () => {
  const store = createStore();
  const mocked = new Map();
  return {
    ...store,
    get: (a) => mocked.has(a) ? mocked.get(a) : store.get(a),
    mockAtom: (a, v) => mocked.set(a, v),
  };
};
```
Hm, it doesn't work well, because values in the original store aren't changed.

**Reply by dephiros** [reply]
@dai-shi , thanks for the reply.
That is too bad. I guess I can play around with `jest.mock` and just sub the original dependencies
